feat(linter): implement `no-negated-condition` rule

baseballyama opened the pull request:
implement: https://eslint.org/docs/latest/rules/no-negated-condition

baseballyama committed:
feat(linter): implement `no-negated-condition` rule

Sysix commented on crates/oxc_linter/src/rules/eslint/no_negated_condition.rs:
```diff
@@ -0,0 +1,127 @@
+                }
+
+                if is_negated_if(if_stmt) {
+                    ctx.diagnostic(no_negated_condition_diagnostic(node.span()));
```
```suggestion
                    ctx.diagnostic(no_negated_condition_diagnostic(if_stmt.span));
```

baseballyama replied:
Updated!: https://github.com/oxc-project/oxc/pull/8149/commits/145014607b69f6e49c7c40a7526595d545683276

Sysix commented on crates/oxc_linter/src/rules/eslint/no_negated_condition.rs:
```diff
@@ -0,0 +1,127 @@
+            }
+            AstKind::ConditionalExpression(conditional_expr) => {
+                if is_negated_if_conditional(conditional_expr) {
+                    ctx.diagnostic(no_negated_condition_diagnostic(node.span()));
```
```suggestion
                    ctx.diagnostic(no_negated_condition_diagnostic(conditional_expr.span));
```

baseballyama replied:
Updated!: https://github.com/oxc-project/oxc/pull/8149/commits/145014607b69f6e49c7c40a7526595d545683276

Sysix replied:
@baseballyama you do not need to call `.span()`. `.span` is enough.
Then you can remove the import `use oxc_span::GetSpan;`. 
Sorry I should have told you what my indentations are.

baseballyama replied:
Ohhhh; I didn't see this carefully🙏
Updated: https://github.com/oxc-project/oxc/pull/8149/commits/155d9ea0fc90f90dc2709b36ceaf4b5af8b044a3

And maybe there are same issue on other PRs, so I will check this.

Sysix commented on crates/oxc_linter/src/rules/eslint/no_negated_condition.rs:
````diff
@@ -0,0 +1,127 @@
+    /// ```
+    NoNegatedCondition,
+    style,
+);
````
```suggestion
    pending,
);
```

We could proably support simple if statements, with `pending` we tell that a auto-fix / suggestion is possible:
https://oxc.rs/docs/guide/usage/linter/rules.html

baseballyama replied:
The original ESLint rule did not implement a fix, so I followed that approach. However, it does seem feasible to implement an automatic fix. I’ll consider to implement this in a another PR.

baseballyama replied:
Updated!: https://github.com/oxc-project/oxc/pull/8149/commits/145014607b69f6e49c7c40a7526595d545683276

Sysix reviewed (commented):
thanks for the work. I commented some possible improvements.  
They are all probably only nitpicking :')

baseballyama committed:
update according to comment

baseballyama committed:
span() -> span

Sysix reviewed (commented):
👍  you can resolve the discussions (I do not have the rights for that ^^)

camc314 reviewed (changes requested):
i think we've already got this rule covered by the unicorn variant?

do you mind porting the test cases to there and seeing if the all pass correctly?

https://github.com/oxc-project/oxc/blob/bd0693bcd01cc826a4717779558e06ce99262732/crates/oxc_linter/src/rules/unicorn/no_negated_condition.rs#L0-L1

baseballyama committed:
merge eslint-plugin-unicorn

baseballyama commented:
@camc314 

I added tests to the unicorn's rule, and they all passed.
I think it would be better to have a rule on the ESLint side as well, so I rewrote it to forward the processing.
What do you think?

baseballyama committed:
fix lint

Sysix commented:
@camc314 personally I like this remapping. Can we do this the same for other rules and maybe fix #8009 ?

camc314 commented:
i haven't looked into https://github.com/oxc-project/oxc/issues/8009.

but we shouldn't have the same code doing the same thing/

That said, it would be OK to map `unicorn/no-negated-condition` to `eslint/no-negated-condition` as we do for some of the TS rules.

To clarify, we would change the following

https://github.com/oxc-project/oxc/blob/3858221f453b8037e5d1d6f5862740db24474345/crates/oxc_linter/src/config/rules.rs#L159-L161

to be something like:

```diff
diff --git a/crates/oxc_linter/src/config/rules.rs b/crates/oxc_linter/src/config/rules.rs
index e661f876f..251026b5f 100644
--- a/crates/oxc_linter/src/config/rules.rs
+++ b/crates/oxc_linter/src/config/rules.rs
@@ -160,6 +160,10 @@ fn transform_rule_and_plugin_name<'a>(
         return (rule_name, "eslint");
     }
 
+    if plugin_name == "unicorn" && rule_name == "no-negated-condition" {
+        return ("no-negated-condition", "eslint");
+    }
+
     (rule_name, plugin_name)
 }
 ```

that way, regardsless of whether you have `eslint/no-negated-condition` OR `no-negated-condition` OR `unicorn/no-negated-condition`, it should turn on the rule correctly.

camc314 commented:
i looked into https://github.com/oxc-project/oxc/issues/8009 briefly and i'm pretty sure it's unrelated to this. it's (possibly - although i couldn't reproduce it) a bug somewhere in our config code

baseballyama committed:
update transform_rule_and_plugin_name

baseballyama commented:
@camc314 

Thank you for your review! (And other PRs also!)

I updated the PR according to your comment.
https://github.com/oxc-project/oxc/pull/8149/commits/24388b5b54110e06e79ca871c1f96251c4a45786

baseballyama committed:
fix

camc314 commented on crates/oxc_linter/src/rules/unicorn/no_negated_condition.rs:
this file needs to be moved into `rules/eslint`

baseballyama replied:
I understand what this process means.
updated!

```rust
    if plugin_name == "unicorn" && rule_name == "no-negated-condition" {
        return ("no-negated-condition", "eslint");
    }
```

camc314 commented on crates/oxc_linter/src/snapshots/unicorn_no_negated_condition.snap:
this one should be deleted

baseballyama replied:
Did!

baseballyama committed:
move from unicorn -> eslint

baseballyama committed:
change test order

camc314 reviewed (approved):
thanks!
